sockeye/inference.py

```diff
@@ -1587,12 +1587,17 @@ def _beam_search(self,
                         beam_histories[sent]["normalized_scores"].append(
                             normalized_scores[rows].asnumpy().flatten().tolist())
 
-            if self.beam_search_stop == C.BEAM_SEARCH_STOP_FIRST and self.batch_size == 1:
-                # TODO: extend to work with batch_size > 1 (i.e., one stopped for each sentence)
-                if mx.nd.sum(finished).asscalar() > 0:
-                    break
+            if self.beam_search_stop == C.BEAM_SEARCH_STOP_FIRST:
+                if self.batch_size == 1:
+                    if finished.sum().asscalar() > 0:
+                        break
+                else:
+                    # if at least one finished hypotheses for each input
+                    at_least_one_finished = finished.reshape((self.batch_size, self.beam_size)).sum(axis=1) > 0
+                    if at_least_one_finished.sum().asscalar() == self.batch_size:
+                        break
             else:
```

Comment:
Isn't `reshape()` basically free? If so we could get rid of the switching on `self.batch_size`, clearing up the code a little bit.

Reply:
Yeah good point. I will simplify and then merge.